### PR TITLE
Catch if parent row exists when we try to add it

### PR DESCRIPTION
**Pull request by tomneedham**
## Description
Sometimes the row that we are trying to insert (in order to use metadata where present) already exists which throws an exception and this cancels the whole repair operation. Instead it is safe to skip this.

## Motivation and Context
Running repair step and getting an exception that the row already exists which stops the repair step.

## How Has This Been Tested?
Manual repair run and unit tests.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tomneedham** [comment]
@pvince81 I need your assistnace here. According to https://github.com/owncloud/core/blob/master/lib/private/Repair/RepairMismatchFileCachePath.php#L169 the parent entry should not exist - but in the wild we have seen a case where it does exist and throws a DB exception. Hints on why it would not show up in this query - and then suddenly exist?

File: lib/private/Repair/RepairMismatchFileCachePath.php [hunk 1]
```diff
@@ -407,7 +408,11 @@ private function getOrCreateEntry($storageId, $path, $reuseFileId = null) {
 			$values['fileid'] = $qb->createNamedParameter($reuseFileId);
 		}
 		$qb->insert('filecache')->values($values);
-		$qb->execute();
+		try {
+			$qb->execute();
+		} catch (UniqueConstraintViolationException $e) {
+			// Skip if the entry already exists
```
Comment (PVince81) [inline lib/private/Repair/RepairMismatchFileCachePath.php]: we should definitely log this, also include info like file id / parent, whatever can help debug

**Comment by PVince81** [comment]
```
09:51:09 TypeError: Argument 3 passed to OC\Repair\RepairMismatchFileCachePath::__construct() must implement interface OCP\ILogger, none given, called in /var/lib/jenkins/workspace/owncloud-core_core_PR-30451-3XCYYOVY2XYXGG2PZLXZCGTPPGRZHI3L6HJCYE3R73WXJGZU7NHQ/lib/private/Repair.php on line 130 and defined in /var/lib/jenkins/workspace/owncloud-core_core_PR-30451-3XCYYOVY2XYXGG2PZLXZCGTPPGRZHI3L6HJCYE3R73WXJGZU7NHQ/lib/private/Repair/RepairMismatchFileCachePath.php:64
```

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by PVince81** [comment]
please backport

**Pull request merged**